rocAL/rocAL/source/coco_file_source_reader.cpp

```diff
@@ -56,19 +64,68 @@ Reader::Status COCOFileSourceReader::initialize(ReaderConfig desc)
     //    std::cout<<"Metadata reader not initialized for COCO file source\n";
 
     ret = subfolder_reading();
+
+    // if _keypoint is set we need to process by annotation. Iterate the map and for every annotation
+    // check if the image is present in the path user shared. If so add it in a new temp list and assign
+    // it to the _file_names list at the end.
+    if(_keypoint)
+    {
+        std::vector<std::string> temp_file_names;
+        for (auto &elem : _annotation_image_key_map)
+        {
+            auto itr = std::find(_files.begin(), _files.end(), elem.second);
+            if(itr != _files.end())
+            {
+                int index = std::distance(_files.begin(), itr);
+                _annotation_ids.push_back(elem.first);
+                temp_file_names.push_back(_file_names[index]);
+            }
+        }
+        _file_names.clear();
+        _file_names = temp_file_names;
+
+        //To handle case where number of images < batch size for multi annotation
+        size_t temp_file_names_size = temp_file_names.size();
+        size_t remaining_images = temp_file_names_size % _batch_count;
+        if (remaining_images > 0)
+        {
+            _in_batch_read_count = remaining_images;
+            _last_file_name = _file_names[temp_file_names_size - 1];
+            std::string last_annotation_id = _annotation_ids[temp_file_names_size - 1];
+            for (uint i = 0; i < (_batch_count - remaining_images); i++)
+                 _annotation_ids.push_back(last_annotation_id);
+            replicate_last_image_to_fill_last_shard();
+        }
+        _file_count_all_shards = _file_id = _file_names.size();
+    }
+
     // the following code is required to make every shard the same size:: required for multi-gpu training
     if (_shard_count > 1 && _batch_count > 1) {
         int _num_batches = _file_names.size()/_batch_count;
-        int max_batches_per_shard = (_file_count_all_shards + _shard_count-1)/_shard_count;
-        max_batches_per_shard = (max_batches_per_shard + _batch_count-1)/_batch_count;
+        int max_batches_per_shard = (_file_count_all_shards + _shard_count - 1)/_shard_count;
+        max_batches_per_shard = (max_batches_per_shard + _batch_count - 1)/_batch_count;
         if (_num_batches < max_batches_per_shard) {
             replicate_last_batch_to_pad_partial_shard();
         }
     }
     //shuffle dataset if set
     _shuffle_time.start();
     if (ret == Reader::Status::OK && _shuffle)
-        std::random_shuffle(_file_names.begin(), _file_names.end());
+    {
+        if(_keypoint)
+        {
+            //shuffle both file names and annotations id's in same order
+            auto seed = unsigned(std::time(0));
```

Comment:
Check if this gives different outputs for different iteration?
Similar to epochs in training run it multiple time after calling reset.

Reply:
Yeah i have checked it. It gives output images in different order for every iteration
since we are using time(0) as the seed which is different for iteration